### PR TITLE
Dynamically query CF API for site bucket

### LINKED ISSUE #2183: Dynamically query CF API for site bucket
#### User Story

When a user creates a new Federalist site, an S3 bucket will be created to publish the site to.

#### Background

Currently, we use an [S3 bucket service bound](https://github.com/18F/federalist/blob/staging/manifest.yml#L15) to the Federalist app to get the credentials the necessary credentials to publish a site to a bucket.  We will need to use the Cloud Foundry (CF) REST API to dynamically create and query S3 buckets related to sites.

#### Acceptance Criteria

- [x] A newly created site will create a new S3 bucket
  - [x] Waiting to update `federalist-proxy` before enabling new bucket creation
- [x] All current sites will continue to use the shared S3 buckets
- [x] Creating a new site will maintain the same workflow
- [x] Bind a `deploy-user` service to support CF API auth see [`federalist-builder`](https://github.com/18F/federalist-builder)
  - [x] Add to both staging and prod manifest
- [ ] Bind `BUILD_SPACE_GUID` (*Change to CF_SPACE_GUID), `CLOUD_FOUNDRY_OAUTH_TOKEN_URL`, and `CLOUD_FOUNDRY_API_HOST` to app environment
  - [x] Add to staging
  - [ ] Add to prod
  - [x]  (*Change to CF_SPACE_GUID)
- [x] Check the permissions needed to delete an S3 cloud foundry service and service keys see [`cfApiClient` delete methods](https://github.com/18F/federalist/pull/2187/files/2c8f0f444f9edd40784b85ad17d53c0a544414fd#diff-db0c560c6ac06bb4554e6a31338f337cR49)
  - [x] Verify permissions and possibly move to a future task
- [x] Come to a consensus on S3 Service naming convention
  - [x] Update naming convention to {owner}-{repo}
- [ ] Verify and add a service key to the staging and production shared S3 Bucket service (new buckets get service keys created on creation) to get AWS credentials for API client to pass to the AWS client.

#### Level of effort

_Medium_

#### Implementation outline

- [x] Bind `federal-user-deploy` service to the Federalist app so we can get the credentials to authenticate to the CF API *see [federalist-builder client auth](https://github.com/18F/federalist-builder/blob/staging/src/cloud-foundry-auth-client.js) for example*

- [x] Use the CF API to create a site specific S3 bucket service *see [federalist-builder client api](https://github.com/18F/federalist-builder/blob/staging/src/cloud-foundry-api-client.js) for implementation reference*

- [x] Migrate the Federalist database to store bucket info
  - One to One site to bucket relationship
  - Save the CF S3 Bucket service name for querying the CF API
    - *ie. `cf_name = <site-name>-<guid>`*
  - Save the AWS S3 bucket name
    - *ie. `s3_name = <bucket name generated by cloud foundry>`*
  - Save the AWS S3 region *ie `s3_region = us-west-2`*
  - The AWS S3 bucket name and region will allow us to know the public bucket url
    - *ie `<s3_bucket_name>.<s3_region>.amazonaws.com` see [federalist-proxy](https://github.com/18F/federalist-proxy/blob/staging/staging_manifest.yml#L9)*

- [x] Insert shared S3 bucket  current sites sitting in the  will insert there information
  - `cf_name = federalist-production-s3`
  - `s3_name = cg-1234567`
  -  `s3_region = us-east-1`

- [x] Modify the relevant services and models to dynamically query CF API for the bucket credentials based on the site instead of using the bound S3 bucket service credentials
  - [x] Update [api/services/S3Helper](https://github.com/18F/federalist/blob/3a43f70cc30e6cdf4946285c548bb3c64d5cd017/api/services/S3Helper.js#L8-L12) to create `s3Client` dynamically based on `site`
    - **Effected Files**
      - [api/services/S3SiteRemover.js](https://github.com/18F/federalist/blob/05bce36f722d967dadbc617769d8ba44f38429fa/api/services/S3SiteRemover.js)
      - [api/services/S3PublishedFileLister.js](https://github.com/18F/federalist/blob/05bce36f722d967dadbc617769d8ba44f38429fa/api/services/S3PublishedFileLister.js)
  - [x] The generation of the bucket will need to happen in the [api/services/SiteCreator.js](https://github.com/18F/federalist/blob/staging/api/services/SiteCreator.js#L171-L186)
  - [x] Update [api/services/SQS.js](https://github.com/18F/federalist/blob/3a43f70cc30e6cdf4946285c548bb3c64d5cd017/api/services/SQS.js#L63-L82) to retrieve `AWS_DEFAULT_REGION`, ` AWS_ACCESS_KEY_ID`, ` AWS_SECRET_ACCESS_KEY_ID`, and `BUCKET` dynamically based on `site`
  - [x] ~Update [api/models/build-log.js](https://github.com/18F/federalist/blob/3a43f70cc30e6cdf4946285c548bb3c64d5cd017/api/models/build-log.js) to remove S3 secrets dynamically~ (Unaffected)

### LINKED ISSUE #2184: Create buckets based on Github user and repositories based on pattern
#### User Story

As a manager of Federalist, I want to create buckets based on Github user and repositories.

#### Acceptance Criteria

- [x] Use a pattern to create bucket service via CloudFoundry.
- [x] Decide if information is stored in database or environment
- [x] Manage binding of services dynamically

### LINKED ISSUE #2185: Change build process to support uploading site to distinct S3 buckets.
Change build process to support uploading site to distinct S3 buckets.

#### Acceptance Criteria

- [ ] All new sites use their own S3 bucket
- [ ] Old sites stay within existing structure
- [ ] There is an option in the UI to "convert" "directory" site into "bucket" site.
- [x] Preserve the path structure of URL, while the subdomain will be distinct for each bucket site
- [x] In Federalist build model callback, the app needs to determine if the current is a directory or S3 and act accordingly
- [ ] Ensure we handle any issues when credentials are not retrieved properly

### LINKED ISSUE #2186: Create a bucket when new site is created
#### User Story

When a federalist user creates a new site, a bucket is created for the site

#### Background (Optional)

Create the option to create a new bucket for a new site

#### Acceptance Criteria

- [x] When a new site is created, enable the ability to create a new bucket
- [x] A new site with the new bucket option creates the new bucket for the site
- [x] A new site without the new bucket option uses the shared bucket

#### Implementation outline (if higher than "low" effort):

- [x] Update the `create site` method to include the creation of a S3 bucket service
- [x] Save the bucket information to the database

### PR DESCRIPTION
**Pull request by apburnes**
## Description

Use the Cloud Foundry REST API to dynamically create and query S3 buckets related to sites.

Closes #2183 
Closes #2184 
Closes #2185 
Closes #2186

## ToDo
- [x] Migrate the Federalist database to store bucket info
  - Adds columns `cfInstanceName`, `awsBucketName`, `awsBucketRegion` to the `Site` model

- [x] Inserts shared S3 bucket information into all of the current sites
  - `cfInstanceName = 'federalist-<environment>-s3'`
  - `awsBucketName = <cf-generated-bucket-name>`
  - `awsBucketRegion = <aws-bucket-region>`

- [x] Add authentication and client to interact with Cloud Foundry API

- [x] The create or use shared bucket service when creating a new site [api/services/SiteCreator.js](https://github.com/18F/federalist/blob/staging/api/services/SiteCreator.js#L171-L186)

- [x] Update [api/services/S3Helper](https://github.com/18F/federalist/blob/3a43f70cc30e6cdf4946285c548bb3c64d5cd017/api/services/S3Helper.js#L8-L12) to create `s3Client` dynamically based on `site`
  - **Effected Files**
    - [api/services/S3SiteRemover.js](https://github.com/18F/federalist/blob/05bce36f722d967dadbc617769d8ba44f38429fa/api/services/S3SiteRemover.js)
    - [api/services/S3PublishedFileLister.js](https://github.com/18F/federalist/blob/05bce36f722d967dadbc617769d8ba44f38429fa/api/services/S3PublishedFileLister.js)

- [x] Update [api/services/SQS.js](https://github.com/18F/federalist/blob/3a43f70cc30e6cdf4946285c548bb3c64d5cd017/api/services/SQS.js#L63-L82) to retrieve `AWS_DEFAULT_REGION`, ` AWS_ACCESS_KEY_ID`, ` AWS_SECRET_ACCESS_KEY_ID`, and `BUCKET` dynamically based on `site`

- [x] ~Update [api/models/build-log.js](https://github.com/18F/federalist/blob/3a43f70cc30e6cdf4946285c548bb3c64d5cd017/api/models/build-log.js) to remove S3 secrets dynamically~ (Unaffected)

### CODE REVIEW

File: api/models/site.js [hunk 1]
```diff
@@ -175,6 +183,18 @@ module.exports = (sequelize, DataTypes) => {
       values: ['active', 'inactive'],
       defaultValue: 'active',
     },
+    cfInstanceName: {
```
Comment (amirbey) [inline api/models/site.js]: what is cfInstanceName?

Comment (amirbey) [reply api/models/site.js]: ok .. from below it appears to be a service instance? could "cfServiceName" be better?

File: api/services/SiteCreator.js [hunk 2]
```diff
@@ -97,13 +99,36 @@ function checkGithubRepository({ user, owner, repository }) {
...
 
+function validateSite(params) {
+  if (params.sharedBucket) {
+    return buildSite(params, config.s3);
+  }
+
+  return apiClient.createSiteBucket(params.repository)
+    .then((response) => {
+      const s3 = {
+        instanceName: params.repository,
```
Comment (amirbey) [inline api/services/SiteCreator.js]: rename cfInstanceName => cfServiceName ?

File: api/utils/cfApiClient.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    ));
+  }
+
+  createSiteBucket(name, keyIdentifier = 'key', serviceName = 'basic-public') {
+    return this.createS3ServiceInstance(name, serviceName)
+      .then(res => this.createServiceKey(name, res.metadata.guid, keyIdentifier));
+  }
+
+
+  // TODO Check Permissions to Delete Services
```
Comment (amirbey) [inline api/utils/cfApiClient.js]: is there an issue or checklist item to come back to this?

File: config/env/production.js [hunk 4]
```diff
@@ -24,6 +24,7 @@ if (s3Creds) {
     secretAccessKey: s3Creds.secret_access_key,
     region: s3Creds.region,
     bucket: s3Creds.bucket,
+    instanceName: s3Creds.instance_name,
```
Comment (amirbey) [inline config/env/production.js]: serviceName?

File: config/env/test.js [hunk 5]
```diff
@@ -13,6 +13,7 @@ module.exports = {
     secretAccessKey: '456def',
     region: 'us-gov-west-1',
     bucket: 's3-bucket',
+    instanceName: 'federalist-dev-s3',
```
Comment (amirbey) [inline config/env/test.js]: serviceName?

File: migrations/20190320200142-add-bucket-info-to-site.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+const insertIfMissing = (db) => {
+  db.runSql(cmdCountRecords)
+    .then(res => res.rows[0].count)
+    .then((count) => {
+      if (count !== '0') db.runSql(cmdUpdateSiteRecords);
+    });
+};
+
+exports.up = (db, callback) => db.addColumn('site', 'cfInstanceName', dataType)
```
Comment (amirbey) [inline migrations/20190320200142-add-bucket-info-to-site.js]: s3ServiceName ???

File: api/models/site.js (hunk 1 above)
Comment (apburnes) [reply api/models/site.js]: Sounds good

Comment (amirbey) [reply api/models/site.js]: or s3ServiceName .. i think is better

File: api/utils/cfApiClient.js (hunk 3 above)
Comment (apburnes) [reply api/utils/cfApiClient.js]: I'll update the PR

File: migrations/20190320200142-add-bucket-info-to-site.js (hunk 6 above)
Comment (apburnes) [reply migrations/20190320200142-add-bucket-info-to-site.js]: LGTM

File: api/models/site.js (hunk 1 above)
Comment (apburnes) [reply api/models/site.js]: Updated

File: api/services/SiteCreator.js (hunk 2 above)
Comment (apburnes) [reply api/services/SiteCreator.js]: Updated

File: api/utils/index.js [hunk 7]
```diff
@@ -5,6 +5,16 @@ const moment = require('moment');
 const config = require('../../config');
 const { logger } = require('../../winston');
 
+function generateS3ServiceName(owner, repository) {
```
Comment (amirbey) [inline api/utils/index.js]: i realized that an owner could have a hyphen and therefore create a naming conflict. for example:

owner1: hello-awesome
repo1: people

vs.

owner2: hello
repo2: awesome-people

bothe create "hello-awesome-people"

Comment (amirbey) [reply api/utils/index.js]: wordy but probably easier to work with ... especially when troubleshooting

Comment (apburnes) [reply api/utils/index.js]: 👍

**Pull request merged**